### ISSUE TITLE
Explore the training procedure and eventually define the edits needed

### ISSUE THREAD

**Issue by gcroci2** [ANNUBS/annubes#19]
As a reference for the code, see PR #16 

### Data in each epoch

`trials = task.generate_trials()` is always the same across one network's training (`rng` is fixed once for each network), being the entire training set. It's used thousands of times (epochs) to train the net. Are we fine with that? If yes, there is no point in generating the same trials for each epoch every time. 

_What do we think it makes more sense to do here?_

### Mini Batch vs Batch Gradient Descent

- Right now, every epoch is trained with very few samples (< 100, `minibatch_size`). Usually, we need many more batches to train during 1 complete epoch.
_Shouldn't we increase the number of samples used for 1 epoch?_
- Here we're using a **Batch Gradient Descent** (not mini-batch as the var name would suggest). In Batch Gradient Descent, all the training data is taken into consideration to take a single step. We take the average of the gradients of all the training examples and then use that mean gradient to update our parameters - default reduction of `nn.MSELoss`. So that’s just one step of gradient descent in one epoch.
_Is this a correct way of training from a ML point of view?_
- **Mini Batch Gradient Descent**: in [DeepRank2](https://github.com/DeepRank/deeprank2), we define a batch_size for our 10k or whatever dataset, and we pass in all the batches within one epoch, averaging the losses across all batches. The ref paper for Thomas' work ([Training Excitatory-Inhibitory Recurrent Neural Networks for Cognitive Tasks: A Simple and Flexible Framework](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004792#sec002)) is using Mini Batch Gradient Descent, even if it's not clear to me how from the code ([GitHub - frsong/pycog: Train excitatory-inhibitory recurrent neural networks for cognitive tasks.](https://github.com/frsong/pycog)).
_Does it make sense to implement something like this instead of what is in place now?_
_Ask Shirin and Thomas for information about the choices made so far._

### Validation

_Any particular reason why validation starts after epoch 200? I think we should properly implement some early stopping callback._

-------------------------

_Useful notes_
Batch size: defines the number of samples to work through before updating the internal model parameters.
Number of epochs: defines the number of times that the learning algorithm will work through the entire training dataset.
[Batch, Mini Batch & Stochastic Gradient Descent](https://towardsdatascience.com/batch-mini-batch-stochastic-gradient-descent-7a62ecba642a)

This issue blocks #15  

We decided to start with the reinforcement part first, in particular by trying to mimic this article procedure: https://elifesciences.org/articles/21492

**Comment by gcroci2** [comment]
### Rationale
- A major goal in neuroscience is to understand the relationship between an animal’s behavior and how this is encoded in the brain. 
- Typical experiment: training an animal to perform a task and recording the activity of its neurons while the animal carries out the task.
- To complement these experimental results, researchers “train” artificial neural networks to simulate the same tasks on a computer.
   - Unlike real brains, artificial neural networks provide complete access to the “neural circuits” responsible for a behavior, offering a way to study and manipulate the behavior in the circuit. 

Supervised learning: the network is given the correct response on each trial in the form of a continuous target output to be followed. 

Reinforcement learning: it provides evaluative feedback to the network on whether each selected action was 'good' or 'bad.'
- Song et al.'s networks consisted of two parts
   - A “decision network” that uses sensory information to select actions that lead to the greatest reward.
   - A “value network” that predicts how rewarding an action will be.

**Comment by gcroci2** [comment]
Issue closed as we have decided to implement similar functionality within the [NeuroGym](https://github.com/neurogym/neurogym) package.
